app/grandchallenge/algorithms/serializers.py

```diff
@@ -219,63 +217,65 @@ def validate(self, data):
-
-        # use interface defaults if no value was provided
-        algorithm_input_pks = {
-            a.pk
-            for a in validated_data["algorithm_image"].algorithm.inputs.all()
-        }
-        input_pks = {i["interface"].pk for i in inputs_data}
-        defaults = validated_data["algorithm_image"].algorithm.inputs.filter(
+        default_inputs = self._algorithm.inputs.filter(
             id__in=list(algorithm_input_pks - input_pks),
             default_value__isnull=False,
         )
-
-        for d in defaults:
-            civ = ComponentInterfaceValue(
-                interface_id=d.id, value=d.default_value
+        # Use default interface values if not present
+        for interface in default_inputs:
+            if interface.default_value:
+                inputs.append(
+                    {"interface": interface, "value": interface.default_value}
+                )
+
+        reformatted_inputs = self.reformat_inputs(serialized_civs=inputs)
+        data.update(reformatted_inputs)
+
+        if Job.objects.get_jobs_with_same_inputs(
+            data=data,
+            algorithm_image=data["algorithm_image"],
+            algorithm_model=data["algorithm_model"],
+        ):
+            raise serializers.ValidationError(
+                "A result for these inputs with the current image "
+                "and model already exists."
             )
-            civ.full_clean()
-            civ.save()
-            component_interface_values.append(civ)
 
-        job = Job.objects.create(
-            **validated_data,
+        return data
+
+    def create(self, validated_data):
+        job = Job.objects.create_job_and_process_inputs(
+            data=validated_data,
             extra_logs_viewer_groups=[
                 validated_data["algorithm_image"].algorithm.editors_group
             ],
-            input_civ_set=component_interface_values,
-            time_limit=validated_data["algorithm_image"].algorithm.time_limit,
-        )
-        job.sort_inputs_and_execute(
-            upload_session_pks=upload_session_pks,
-            user_upload_pks=user_upload_pks,
         )
-
         return job
+
+    def reformat_inputs(self, *, serialized_civs):
+        """
+        Takes serialized CIV data and turns it into a dictionary:
+        {
+            "interface_slug_1": "value_1",
+            "interface_slug_2": "value_2"
+        }
+        This representation of CIV data corresponds to how CIV data is stored in
+        a form's cleaned_data.
+        """
+        possible_keys = [
+            "image",
+            "value",
+            "file",
+            "user_upload",
+            "upload_session",
+        ]
+        data = {}
+        for civ in serialized_civs:
+            interface_slug = civ["interface"].slug
+            for key in possible_keys:
+                if key in civ:
+                    data[interface_slug] = civ[key]
+                    break
```

Comment:
The user should be informed with a 400 error should occur if there is no match, or multiple matches, right?

Reply:
That type of validation is already done by the ComponentInterfaceValuePostSerializer, so the data we get here is guaranteed to have one and only one of the possible keys.

Reply:
You are relying on how things are organised today. Every method should be robust. You can future proof yourself by explicitly handling the unexpected cases. Maybe you don't need to send something to the user here but at the very least raise an exception when your assumption breaks down.